Introduce BanMan

theuni opened the pull request:
BanMan is not a bad Man, he's just a different class of Man. He's an ex-ConnMan Man, looking to take a different path and make a new namespace for himself.

Beware BanMan's banhammer as he banishes bandits with the bandwidth to jam him. Not standing for non-canon, he’ll ban and he’ll ban. Some heroes wear capes, but BanMan? A Bandana.

--

Despite the diff size, this is mostly move-only. It breaks the ban/unban functions out of CConnman and into a new class because, while logically bans are tied to connections, they're really just entries in a database. Like CConnman, a global is still required due to RPC (and qt). I plan to address this along with CAddrMan, which I'll be breaking out next.

This also makes testing easier as different implementations can be dropped in.

There are a few small behavioral changes here, which are pretty insignificant:
- Banning no longer implies disconnecting. If you want both, you need to call both.
- For simplicity, The ban db is read in the constructor, meaning that it happens in init rather than net.
- RPC returns RPC_DATABASE_ERROR if the bandb is not loaded. This should not be possible now, but the idea is for future changes to allow us to disable p2p but still interact with the ban commands.

theuni committed:
net: Break disconnecting out of Ban()

These are separate events which need to be carried out by separate subsystems.

This also cleans up some whitespace and tabs in qt to avoid getting flagged by
the linter.

Current behavior is preserved.

theuni committed:
tests: remove member connman/peerLogic in TestingSetup

theuni committed:
net: split up addresses/ban dumps in preparation for moving them

theuni committed:
banman: create and split out banman

Some say he has always been.

theuni committed:
banman: pass the banfile path in

There's no need to hard-code the path here. Passing it in means that there are
no ordering concerns wrt establishing the datadir.

theuni committed:
banman: pass in default ban time as a parameter

Removes the dependency on arg parsing.

theuni committed:
net: move BanMan to its own files

theuni committed:
scripted-diff: batch-rename BanMan members

-BEGIN VERIFY SCRIPT-
sed -i "s/clientInterface/m_client_interface/g" src/banman.h src/banman.cpp
sed -i "s/setBannedIsDirty/m_is_dirty/g" src/banman.h src/banman.cpp
sed -i "s/cs_setBanned/m_cs_banned/g" src/banman.h src/banman.cpp
sed -i "s/setBanned/m_banned/g" src/banman.h src/banman.cpp
-END VERIFY SCRIPT-

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,65 @@
+
+    banmap_t m_banned;
+    CCriticalSection m_cs_banned;
+    bool m_is_dirty;
```
What about adding `GUARDED_BY(…)` (see #10866 and #11226) annotations here?

Like this:

```
    CCriticalSection m_cs_banned;
    banmap_t m_banned GUARDED_BY(m_cs_banned);
    bool m_is_dirty GUARDED_BY(m_cs_banned);
```

Nit: The lock was renamed from `cs_setBanned` to `m_cs_banned`. All `CCriticalSection`:s in the code base except `m_cs_callbacks_pending` are named without the `m_` prefix. Is `m_cs_` the recommended prefix going forward? More specifically, what prefix is recommended for my work in #10866? :-)  I think there is a point in having consistent prefixes in order to achieve easy grepping for locks.

theuni replied:
See https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md:
> Class member variables have a m_ prefix.

So yes, m_cs_foo is the recommended syntax going forward for member mutexes, I suppose.

theuni replied:
>What about adding GUARDED_BY(…) (see #10866 and #11226) annotations here?

Will do. Additionally, this actually needs another mutex for the db. I've been debating whether it should be fixed as part of this PR, but event-driven db flushes can actually collide with the one on a timer. addrman has the same issue, and that may be the cause of some addrman corruption we've seen lately (#11252, #11409, #11454).

practicalswift commented:
Concept ACK modulo `GUARDED_BY(…)` addition :-)

JeremyRubin commented:
Concept Ack!

jimpo commented on src/addrdb.h:
```diff
@@ -94,7 +94,7 @@ class CBanDB
     fs::path pathBanlist;
 public:
-    CBanDB();
+    CBanDB(fs::path pathBanlist);
```
Make explicit?

theuni replied:
Will do.

jimpo commented on src/banman.h:
```diff
@@ -0,0 +1,65 @@
+class BanMan
+{
+public:
+    // Denial-of-service detection/prevention
```
Should this be a class-level comment? (Like placed right above `class BanMan`)

theuni replied:
Sure

jimpo replied:
BanMan still needs class-level comment.

jimpo reviewed (commented):
Concept ACK

TheBlueMatt reviewed (commented):
Looking godd, will give it a full review when it quiets down.

theuni commented on src/rpc/net.cpp:
```diff
@@ -541,7 +541,13 @@ UniValue setban(const JSONRPCRequest& request)
 
-        isSubnet ? g_connman->Ban(subNet, BanReasonManuallyAdded, banTime, absolute) : g_connman->Ban(netAddr, BanReasonManuallyAdded, banTime, absolute);
+        if (isSubnet) {
+            g_connman->DisconnectNode(subNet);
```
Yes, good point. Will fix.

promag replied:
Inverting opens another dumb case: ban -> (unban -> connect) -> disconnect 🙄 Do you think `DisconnectNode()` should be called inside `Ban()` while the lock is held?

theuni replied:
I don't see how this could reasonably happen. Anyway, it's not a departure from the current behavior.

theuni committed:
banman: add thread annotations and mark members const where possible

Also remove misleading comment. ClearBanned is used by rpc as well.

promag commented on src/test/DoS_tests.cpp:
```diff
@@ -44,6 +44,8 @@ BOOST_FIXTURE_TEST_SUITE(DoS_tests, TestingSetup)
 
 BOOST_AUTO_TEST_CASE(DoS_banning)
 {
+    auto connman = std::unique_ptr<CConnman>(new CConnman(0x1337, 0x1337));
```
Use `std::make_unique`.

promag reviewed (commented):
Partial review. Concept ACK.

theuni commented on src/init.cpp:
```diff
@@ -1267,11 +1273,13 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     // need to reindex later.
 
+    assert(!g_banman);
+    g_banman = std::unique_ptr<BanMan>(new BanMan(GetDataDir() / "banlist.dat", &uiInterface, DEFAULT_MISBEHAVING_BANTIME));
```
that's c++14 :(

practicalswift replied:
A substitute for C++14 `std::make_unique` in the form of `MakeUnique` is added to `util.h` as part of #11043 (see https://github.com/bitcoin/bitcoin/pull/11043/commits/27a54eacc07325e9a10cbea71ab03d87a96679b7) which will hopefully be merged soon :-)

jonasschnelli commented:
Nice!
Concept ACK,.. will review soon.

Sjors commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time);
+    void Ban(const CNetAddr& netAddr, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
+    void Ban(const CSubNet& subNet, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
+    void ClearBanned(); // needed for unit testing
```
Would it make sense to move `ClearBanned()` to a subclass that's only used by the test suite?

theuni replied:
This is a misleading comment that needs to be dropped. This is used by the ```clearbanned``` rpc as well.

theuni commented:
Rebased and addressed most of the feedback here.

sdaftuar commented on src/init.cpp:
```diff
@@ -1272,7 +1272,7 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
 
     assert(!g_banman);
-    g_banman = std::unique_ptr<BanMan>(new BanMan(GetDataDir() / "banlist.dat", &uiInterface));
+    g_banman = std::unique_ptr<BanMan>(new BanMan(GetDataDir() / "banlist.dat", &uiInterface, DEFAULT_MISBEHAVING_BANTIME));
```
I think we need to pass `gArgs.GetArg("-bantime", DEFAULT_MISBEHAVING_BANTIME)` here?  Or else I'm missing how the command line argument gets used after this commit.

theuni replied:
Very much so, thanks for catching that.

sdaftuar reviewed (commented):
Concept ACK.  Left a few comments/questions, but looks pretty good.  I still need to verify the move-only-ness of a758cebfc66e0a694506291d925f2ef156661654.

theuni commented on src/init.cpp:
```diff
@@ -1266,11 +1271,13 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     CConnman& connman = *g_connman;
 
-    peerLogic.reset(new PeerLogicValidation(&connman));
+    peerLogic.reset(new PeerLogicValidation(g_connman.get(), g_banman.get()));
```
Because these aren't intended to be actual globals. We just don't (yet) have the infrastructure to pass them into RPC and qt. Once banman/addrman are split out, I'll be PRing a set of changes that make RPC instantiated. With that done, the ui (rpc/qt/whatever else) will have these interfaces passed in, and they can assume that the pointers are good for their lifetime.

theuni commented on src/init.cpp:
```diff
@@ -1722,5 +1730,11 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     StartWallets(scheduler);
 #endif
 
+    if (g_banman) {
```
My original thinking was that if we add a -nop2p option (which I'm hoping we can do soon), the bandb would be disabled as well. But thinking about it now, one doesn't exclude the other. So yes, an assert makes more sense here. Will do.

theuni commented on src/net.cpp:
```diff
@@ -38,8 +38,11 @@
+static const int DUMP_PEERS_INTERVAL = 60 * 15;
+
+// Dump addresses to banlist.dat every 15 minutes (900s)
+static const int DUMP_BANS_INTERVAL = 60 * 15;
```
Yep, will change.

theuni commented on src/net.cpp:
```diff
@@ -2405,7 +2413,9 @@ void CConnman::Stop()
         DumpAddresses();
-        DumpBanlist();
+        if (m_banman) {
+            m_banman->DumpBanlist();
```
Good point, will remove.

theuni commented on src/qt/rpcconsole.cpp:
```diff
@@ -1155,16 +1155,17 @@ void RPCConsole::banSelectedNode(int bantime)
+        const CNodeCombinedStats *stats = clientModel->getPeerTableModel()->getNodeStats(detailNodeRow);
+        if(stats) {
+            g_connman->Ban(stats->nodeStats.addr, BanReasonManuallyAdded, bantime);
+            g_connman->DisconnectNode(id);
```
When accepting an ip as a string from the user I agree. But in this case the request is for a specific node, so I would argue that the id is more in line with what the user would expect.

I don't care hugely either way though.

As an aside, what we really need here is kick/ban/kick+ban options. But I don't think it's worth adding the complexity?

theuni commented on src/qt/rpcconsole.cpp:
```diff
@@ -1163,8 +1163,12 @@ void RPCConsole::banSelectedNode(int bantime)
+                g_banman->Ban(stats->nodeStats.addr, BanReasonManuallyAdded, bantime);
+            }
+            if (g_connman) {
+                g_connman->DisconnectNode(id);
```
Same response :)

Note that these comments are a bit stale. We used to track down the entry by looking through all node addresses, but now we use the id directly. I'll update those comments independently.

Also, see the behavior of disconnectSelectedNode above. I think it makes sense to be consistent with that?

For the sake of not over-complicating this, It may make sense to just keep the old behavior for this PR and switch to id in a follow-up.

theuni commented on src/rpc/net.cpp:
```diff
@@ -509,8 +509,8 @@ UniValue setban(const JSONRPCRequest& request)
-    if(!g_connman)
-        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    if(!g_banman)
+        throw JSONRPCError(RPC_DATABASE_ERROR, "Error: Ban database not loaded");
```
I didn't bother with docs because as far as I know, it's been impossible to receive this error in practice because g_connman always exists. It won't be until we have a -nop2p that it's relevant.

theuni commented on src/rpc/net.cpp:
```diff
@@ -570,11 +574,11 @@ UniValue listbanned(const JSONRPCRequest& request)
-    if(!g_connman)
-        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    if(!g_banman)
+        throw JSONRPCError(RPC_DATABASE_ERROR, "Error: Ban database not loaded");
```
Same as above. These could be asserts for now, but imo we're less likely to forget about these if errors are hooked up in advance.

theuni commented:
Addressed @sdaftuar's feedback. I went ahead and changed the Disconnect() in qt/rpcconsole back to using IP rather than nodeid to avoid adding unnecessary behavioral changes in a PR that should otherwise be a code refactor.

TheBlueMatt commented:
Concept ACK, needs rebase.

theuni commented:
rebased.

promag commented on src/net.cpp:
```diff
@@ -2559,6 +2553,25 @@ bool CConnman::DisconnectNode(const std::string& strNode)
     return false;
 }
+
+bool CConnman::DisconnectNode(const CSubNet& subnet)
```
In commit `net: Break disconnecting out of Ban()`,

Why `bool` if the return value is not used?

theuni replied:
Just to be consistent with the others.

promag commented on src/qt/rpcconsole.cpp:
```diff
@@ -1143,16 +1143,17 @@ void RPCConsole::banSelectedNode(int bantime)
-	}
+        // Get currently selected peer address
+        int detailNodeRow = clientModel->getPeerTableModel()->getRowByNodeId(id);
+        if(detailNodeRow < 0)
```
In commit `net: Break disconnecting out of Ban()`,

Nit, add space.

theuni replied:
Fixed

promag reviewed (commented):
In commit `net: Break disconnecting out of Ban()`,

Missing call `DisconnectNode()` after:

https://github.com/bitcoin/bitcoin/blob/3bdf242fc68a8d767932c6214455d4d413effbc9/src/net_processing.cpp#L2844

theuni commented:
> In commit net: Break disconnecting out of Ban(),
> 
> Missing call DisconnectNode() after:

This is handled by setting fDisconnect directly above.

TheBlueMatt commented:
Heh, rebase needed for the include change.

laanwj commented:
I've added this to high-priority for review, would be nice to move this forward and this blocks further work in this direction.

theuni commented:
@laanwj Thanks!

TheBlueMatt commented:
Needs rebase.

TheBlueMatt commented on src/rpc/net.cpp:
```diff
@@ -544,16 +544,20 @@ UniValue setban(const JSONRPCRequest& request)
-            g_connman->Ban(subNet, BanReasonManuallyAdded, banTime, absolute);
-            g_connman->DisconnectNode(subNet);
+            g_banman->Ban(subNet, BanReasonManuallyAdded, banTime, absolute);
+            if (g_connman) {
```
Is it really worth checking? We already check for g_banman further up, are we really anticipating having a g_banman but not a g_connman?

theuni replied:
Yea, I think it may be useful in the future to run with -p2p=0 in order to view/cleanup your bandb.

TheBlueMatt commented on src/net.h:
```diff
@@ -132,7 +132,7 @@ class BanMan
     // new code.
     ~BanMan();
-    BanMan(fs::path ban_file, CClientUIInterface* client_interface);
+    BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time);
```
I cant say I really understand the desire to avoid the argparsing dep, when you include <util.h> either way. Seems like useless indirection.

theuni replied:
So that we can test independent of the cmdline args.

TheBlueMatt reviewed (commented):
Nice.

theuni commented on src/net.cpp:
```diff
@@ -556,13 +556,7 @@ void CConnman::Ban(const CSubNet& subNet, const BanReason &banReason, int64_t ba
-                pnode->fDisconnect = true;
-        }
-    }
+
```
Good point. Disconnecting the single node seems more correct to me, but I'll revert to the current behavior here to avoid the change.

theuni commented on src/net.cpp:
```diff
@@ -2283,24 +2279,8 @@ bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
-        SetBannedSetDirty(true); // force write
-        DumpBanlist();
-    }
+    nStart = GetTimeMillis();
```
Yes, will remove.

theuni commented on src/init.cpp:
```diff
@@ -1726,5 +1734,9 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     StartWallets(scheduler);
 #endif
 
+    scheduler.scheduleEvery([]{
```
Well the schedule in this case is bitcoind specific. There's no reason for BanMan to know about it, and it's not exclusive to net. Throwing in an app-specific schedule would kinda defeat the purpose of breaking it out.

TheBlueMatt replied:
Ugh, I mean the other annoying bit is that init.o is, by far, our largest compilation module. Any excuse to pull stuff out of init and put knowledge of a module in that module instead of init seems like the right direction to me. Also because we're telling the scheduler to schedule calling a function in that module based on a timer set in that module's header just seems...weird to me. If we want to ensure its easy to add unit tests later you can easily make the scheduler an optional arg and skip dumping if one isn't provided to the constructor.

theuni replied:
@ryanofsky BanMan doesn't rely on a scheduler to function any more than the wallet does, that's a bitcoind thing. Banman just loads/unloads the bandb when it's told.

Imagine a small util to dump/edit the bandb, (which would be completely reasonable). The scheduler would just be a needless dependency.

@TheBlueMatt It's time to create an app context class (and .h/cpp) anyway, that holds these types of objects for the life of bitcoind. I figured one of us would get around to that once enough stuff had piled up in init.cpp :p .

theuni commented:
Addressed @TheBlueMatt's review and squashed. Diff from before (minus the rebase to master):
```patch
diff --git a/src/net.cpp b/src/net.cpp
index 78c08b1..46e7783 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2125,4 +2125,2 @@ bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
 
-    nStart = GetTimeMillis();
-
     uiInterface.InitMessage(_("Starting network threads..."));
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 2981413..e8d731c 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2852,3 +2852,2 @@ bool PeerLogicValidation::SendRejectsAndCheckIfBanned(CNode* pnode)
         else {
-            pnode->fDisconnect = true;
             if (pnode->addr.IsLocal())
@@ -2861,2 +2860,4 @@ bool PeerLogicValidation::SendRejectsAndCheckIfBanned(CNode* pnode)
             }
+            // Disconnect all nodes with this address
+            connman->DisconnectNode(pnode->addr);
         }

```

jimpo commented on src/net_processing.cpp:
```diff
@@ -2853,7 +2853,9 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman)
             else
             {
-                connman->Ban(pnode->addr, BanReasonNodeMisbehaving);
+                if (m_banman) {
```
Commit [banman: create and split out banman].

I feel like this should be checked outside the surrounding `if (pnode->addr.IsLocal())` statement. Doesn't make sense to log "not banning local peer" if it wouldn't have been banned anyway.

theuni replied:
good point, will fix.

jimpo commented on src/qt/rpcconsole.cpp:
```diff
@@ -1200,9 +1204,9 @@ void RPCConsole::unbanSelectedNode()
 
         LookupSubNet(strNode.toStdString().c_str(), possibleSubnet);
-        if (possibleSubnet.IsValid() && g_connman)
+        if (possibleSubnet.IsValid() && g_banman)
```
Commit [banman: create and split out banman].

`g_banman` is guaranteed to exist given the check at the top of the function.

theuni replied:
Yep, will fix.

jimpo commented on src/rpc/net.cpp:
```diff
@@ -544,16 +544,20 @@ UniValue setban(const JSONRPCRequest& request)
-            g_connman->Ban(netAddr, BanReasonManuallyAdded, banTime, absolute);
-            g_connman->DisconnectNode(netAddr);
+            g_banman->Ban(netAddr, BanReasonManuallyAdded, banTime, absolute);
+            if (g_connman) {
```
Commit [banman: create and split out banman].

Perhaps pull the common `DisconnectNode` code out of this `if/else` statement?

theuni replied:
One is a subnet, the other is an address. CSubNet is a subclass of CNetAddr, so we could potentially store a generic one, but imo duplicating the disconnect is more straightforward.

ryanofsky reviewed (commented):
utACK 8c91ffe777ffcd2724f576490068abcbb95d60c0

theuni commented on src/net.cpp:
```diff
@@ -2556,6 +2550,25 @@ bool CConnman::DisconnectNode(const std::string& strNode)
+    bool disconnected = false;
+    LOCK(cs_vNodes);
+    for (CNode* pnode : vNodes) {
+        if (subnet.Match((CNetAddr)pnode->addr)) {
```
sure

theuni commented on src/net_processing.cpp:
```diff
@@ -2849,13 +2849,14 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman)
                 connman->Ban(pnode->addr, BanReasonNodeMisbehaving);
             }
+            // Disconnect all nodes with this address
+            connman->DisconnectNode(pnode->addr);
```
I'm not sure what the behavior change would be, so I guess not :)

theuni commented on src/test/DoS_tests.cpp:
```diff
@@ -54,6 +68,9 @@ BOOST_FIXTURE_TEST_SUITE(DoS_tests, TestingSetup)
 // work.
 BOOST_AUTO_TEST_CASE(outbound_slow_chain_eviction)
 {
+    auto connman = std::unique_ptr<CConnman>(new CConnman(0x1337, 0x1337));
```
will do

mizhang commented on src/net.cpp:
```diff
@@ -2556,6 +2550,25 @@ bool CConnman::DisconnectNode(const std::string& strNode)
+    for (CNode* pnode : vNodes) {
+        if (subnet.Match((CNetAddr)pnode->addr)) {
+            pnode->fDisconnect = true;
+            disconnected = true;
```
if (disconnected == false) {
    disconnected = true;
}      

To add a little bit of performance

laanwj commented:
utACK 0950d72

Sjors commented:
Did some light testing on macOS banning and unbanning unsuspecting testnet nodes. Didn't get in trouble.

theuni commented:
rebased. I just shoved g_banman into NodeImpl for now, but obviously we'll want it broken out as a next step.

jimpo commented:
utACK 521ebf3
